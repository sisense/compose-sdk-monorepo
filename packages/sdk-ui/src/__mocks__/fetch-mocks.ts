import { MockParams } from 'vitest-fetch-mock';
import mockGlobals from './data/mock-globals.json';
import mockPalettes from './data/mock-palettes.json';
import mockWidgetDrilldown from './data/mock-widget-drilldown.json';
import mockJaqlDrilldown from './data/mock-jaql-drilldown.json';
import mockJaqlDates from './data/mock-jaql-dates.json';

export const mockUrl = `http://10.50.0.1:30845`;
export const mockToken = 'fake-token';
export const mockDashboardId = 'fake-dashboard-id';
export const mockWidgetId = 'fake-widget-id';

export const fetchMocks: Record<string, [string, MockParams]> = {
  globals: [JSON.stringify(mockGlobals), { status: 200 }],
  palettes: [JSON.stringify(mockPalettes), { status: 200 }],
  widgetDrilldown: [JSON.stringify(mockWidgetDrilldown), { status: 200 }],
  jaqlDrilldown: [JSON.stringify(mockJaqlDrilldown), { status: 200 }],
  jaqlDates: [JSON.stringify(mockJaqlDates), { status: 200 }],
};
