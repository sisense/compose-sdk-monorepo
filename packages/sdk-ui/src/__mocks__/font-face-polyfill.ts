// Polifill for FontFace class to avoid error in tests
// https://developer.mozilla.org/en-US/docs/Web/API/FontFace

export class FontFaceMock {
  family: string;

  source: string;

  descriptors: unknown;

  constructor(family: string, source: string, descriptors?: unknown) {
    this.family = family;
    this.source = source;
    this.descriptors = descriptors;
  }

  load(): Promise<FontFaceMock> {
    return Promise.resolve(this);
  }
}

globalThis.FontFace = FontFaceMock as never;
