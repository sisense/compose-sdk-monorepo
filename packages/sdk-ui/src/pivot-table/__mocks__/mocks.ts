import * as DM from '@/__test-helpers__/sample-ecommerce';
import { filterFactory, measureFactory } from '@sisense/sdk-data';
import { PivotTableProps } from '../../props';

export const mockPivotTableProps: PivotTableProps = {
  dataSet: DM.DataSource,
  dataOptions: {
    rows: [
      {
        column: DM.Category.Category,
        includeSubTotals: true,
        sortType: 'sortAsc',
      },
      {
        column: DM.Commerce.AgeRange,
        includeSubTotals: true,
        sortType: {
          direction: 'sortDesc',
          by: {
            valuesIndex: 0,
            columnsMembersPath: ['Male'],
          },
        },
      },
      DM.Commerce.Condition,
    ],
    columns: [{ column: DM.Commerce.Gender, includeSubTotals: true }],
    values: [
      {
        column: measureFactory.sum(DM.Commerce.Cost, 'Total Cost'),
        dataBars: true,
        totalsCalculation: 'sum',
      },
      {
        column: measureFactory.sum(DM.Commerce.Revenue, 'Total Revenue'),
        totalsCalculation: 'sum',
      },
    ],
    grandTotals: { title: 'Grand Total', rows: true, columns: true },
  },
  filters: [filterFactory.members(DM.Commerce.Gender, ['Female', 'Male'])],
};
