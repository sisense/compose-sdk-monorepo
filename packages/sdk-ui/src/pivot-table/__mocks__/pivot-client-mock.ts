/* eslint-disable no-unused-vars */
/* eslint-disable sonarjs/no-identical-functions */
import React from 'react';
import { PivotClient } from '@ethings-os/sdk-pivot-client';

// Mock of PivotBuilder class
class MockPivotBuilder {
  private listeners: Record<string, Array<(...args: any[]) => void>> = {};

  on(eventName: string, callback: (...args: any[]) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(callback);
  }

  off(eventName: string, callback: (...args: any[]) => void) {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((cb) => cb !== callback);
    }
  }

  emit(eventName: string, ...args: any[]) {
    if (this.listeners[eventName]) {
      this.listeners[eventName].forEach((callback) => callback(...args));
    }
  }

  render(props: any) {
    // Mock render method - does not actually render anything
    // Simulate emitting a proper React element instead of a string
    const mockElement = React.createElement(
      'div',
      { 'data-testid': 'mocked-pivot-table' },
      'Mocked Pivot Table',
    );
    setTimeout(() => {
      this.emit('pivotElementChange', mockElement);
    }, 0);
  }

  updateJaql(jaql: any) {
    // Mock updateJaql method
    setTimeout(() => {
      this.emit('queryStart');
      // Simulate query end with empty metadata
      setTimeout(() => {
        this.emit('queryEnd', { cellsMetadata: null });
      }, 10);
    }, 0);
  }

  updateDataService(dataService: any) {
    // Mock updateDataService method
  }

  destroy() {
    // Mock destroy method
    this.listeners = {};
  }
}

// Mock of DataService class
class MockDataService {
  private listeners: Record<string, Array<(...args: any[]) => void>> = {};

  on(eventName: string, callback: (...args: any[]) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(callback);
  }

  off(eventName: string, callback: (...args: any[]) => void) {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((cb) => cb !== callback);
    }
  }

  destroy() {
    // Mock destroy method
    this.listeners = {};
  }
}

// Create a mocked PivotClient
export const createMockPivotClient = () => {
  return {
    preparePivotBuilder: vi.fn(() => new MockPivotBuilder()),
    prepareDataService: vi.fn(() => new MockDataService()),
    queryData: vi.fn(),
    socketBuilder: {
      socket: {},
    },
  } as unknown as PivotClient;
};
