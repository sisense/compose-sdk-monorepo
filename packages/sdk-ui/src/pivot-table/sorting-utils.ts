import { JaqlSortDirection, getSortType } from '@sisense/sdk-data';
import { type SortingSettingsChangePayload } from '@sisense/sdk-pivot-client';
import isEqual from 'lodash-es/isEqual';
import isString from 'lodash-es/isString';
import { PivotTableDataOptionsInternal } from '..';
import type { SortDirection, PivotRowsSort } from '../types';

export type PivotRowsSortCriteria = {
  rowIndex: number;
  sort: PivotRowsSort | null;
};

export type PivotSortCriteria = {
  rows: PivotRowsSortCriteria[];
};

export function normalizePivotSort(sortType: SortDirection | PivotRowsSort) {
  return isString(sortType) ? { direction: sortType } : sortType;
}

function getSortBy(
  sortDetails: SortingSettingsChangePayload['sortDetails'],
  dataOptionsInternal: PivotTableDataOptionsInternal,
) {
  return {
    valuesIndex:
      sortDetails.field! -
      (dataOptionsInternal.rows || []).length -
      (dataOptionsInternal.columns || []).length,
    columnsMembersPath: Object.values(sortDetails.measurePath || {}),
  };
}

export function preparePivotRowsSortCriteriaList(
  { type, settings, sortDetails }: SortingSettingsChangePayload,
  dataOptionsInternal: PivotTableDataOptionsInternal,
): PivotRowsSortCriteria[] {
  if (type === 'simple') {
    const isSortedByMeasure = sortDetails.field! >= (dataOptionsInternal.rows || []).length;

    if (isSortedByMeasure) {
      return [
        {
          rowIndex: 0,
          sort: {
            direction: getSortType(settings[0].direction as JaqlSortDirection),
            by: getSortBy(sortDetails, dataOptionsInternal),
          },
        },
      ];
    }

    return [
      {
        rowIndex: sortDetails.field!,
        sort: {
          direction: getSortType(settings[0].direction as JaqlSortDirection),
        },
      },
    ];
  }

  // 'complex' sorting type
  return settings.reduce((sortCriteriaList: PivotRowsSortCriteria[], sortOptions, index) => {
    const rowDataOption = dataOptionsInternal.rows?.[index];
    const existingRowSort = rowDataOption?.sortType && normalizePivotSort(rowDataOption?.sortType);
    const newSort = {
      direction: getSortType(sortOptions.direction as JaqlSortDirection),
      by: getSortBy(sortDetails, dataOptionsInternal),
    };

    const shouldDisableRowSort = !sortOptions.selected && isEqual(existingRowSort?.by, newSort.by);

    if (shouldDisableRowSort) {
      sortCriteriaList.push({
        rowIndex: index,
        sort: null,
      });
    }

    if (sortOptions.selected) {
      sortCriteriaList.push({
        rowIndex: index,
        sort: newSort,
      });
    }

    return sortCriteriaList;
  }, []);
}
