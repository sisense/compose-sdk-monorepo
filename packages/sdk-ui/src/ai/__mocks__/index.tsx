import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import { mockToken, mockUrl } from '@/__mocks__/msw';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';
import AiContextProvider from '../ai-context-provider';

/**
 * A component that wraps an AI hook/component with all required providers.
 * For use in unit tests only.
 *
 * @internal
 */
export const AiTestWrapper = ({ children }: { children: ReactNode }) => (
  <SisenseContextProvider url={mockUrl} token={mockToken}>
    <AiContextProvider>
      <QueryClientProvider
        client={
          new QueryClient({
            defaultOptions: {
              queries: {
                retry: false,
              },
            },
          })
        }
      >
        {children}
      </QueryClientProvider>
    </AiContextProvider>
  </SisenseContextProvider>
);
