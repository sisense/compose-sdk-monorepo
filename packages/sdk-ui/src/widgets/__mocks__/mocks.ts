import { QueryResultData } from '@sisense/sdk-data';

import * as DM from '../../__test-helpers__/sample-ecommerce';
import { mockPivotTableProps } from '../../pivot-table/__mocks__/mocks';
import { ChartWidgetProps, PivotTableWidgetProps } from '../../props';

export const mockPivotTableWidgetProps: PivotTableWidgetProps = {
  ...mockPivotTableProps,
  dataSource: DM.DataSource,
  title: 'Pivot table widget title',
  description: 'Pivot table widget desription',
  styleOptions: {
    width: 800,
    membersColor: true,
    headersColor: true,
    border: true,
    borderColor: 'lightgrey',
  },
} as PivotTableWidgetProps;

export const mockChartWidgetPropsForTable: ChartWidgetProps = {
  chartType: 'table',
  dataSource: DM.DataSource,
  title: 'Chart widget title',
  description: 'Chart widget desription',
  dataOptions: {
    columns: [DM.Commerce.AgeRange, DM.Commerce.Revenue],
  },
};

export const mockResolvedQuery: QueryResultData = {
  columns: [
    {
      name: 'Age Range',
      type: 'text',
    },
    {
      name: 'Revenue',
      type: 'number',
    },
  ],
  rows: [
    [
      {
        data: '19-24',
        text: '19-24',
        blur: false,
      },
      {
        data: 43.98411178588867,
        text: '43.9841117858887',
        blur: false,
      },
    ],
    [
      {
        data: '65+',
        text: '65+',
        blur: false,
      },
      {
        data: 62.803104400634766,
        text: '62.8031044006348',
        blur: false,
      },
    ],
  ],
};
