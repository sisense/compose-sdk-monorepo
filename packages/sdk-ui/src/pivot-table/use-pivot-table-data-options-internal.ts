import { useState } from 'react';
import {
  PivotTableDataOptions,
  PivotTableDataOptionsInternal,
} from '../chart-data-options/types.js';
import { translatePivotTableDataOptions } from '../chart-data-options/translate-data-options.js';
import { PivotSortCriteria } from './sorting-utils.js';
import { safeCombine } from '@/chart-data-options/utils.js';

type UsePivotTableDataOptionsInternalResult = {
  dataOptionsInternal: PivotTableDataOptionsInternal;
  updateSort: (sortCriteria: PivotSortCriteria) => void;
};

/**
 * A hook that returns the JAQL request object from pivot table props.
 *
 * @internal
 */
export const usePivotTableDataOptionsInternal = ({
  dataOptions,
}: {
  dataOptions: PivotTableDataOptions;
}): UsePivotTableDataOptionsInternalResult => {
  const [dataOptionsInternal, setDataOptionsInternal] = useState<PivotTableDataOptionsInternal>(
    translatePivotTableDataOptions(dataOptions),
  );

  return {
    dataOptionsInternal,
    updateSort(sortCriteria: PivotSortCriteria) {
      setDataOptionsInternal({
        ...dataOptionsInternal,
        rows: dataOptionsInternal.rows?.map((row, index) => {
          const rowSortCriteria = sortCriteria.rows.find(({ rowIndex }) => rowIndex === index);

          if (!rowSortCriteria) {
            return row;
          }

          // remove existing row sorting if it was explicitly removed
          if (rowSortCriteria.sort === null) {
            return safeCombine(row, { sortType: undefined });
          }

          return safeCombine(row, { sortType: rowSortCriteria.sort });
        }),
      });
    },
  };
};
