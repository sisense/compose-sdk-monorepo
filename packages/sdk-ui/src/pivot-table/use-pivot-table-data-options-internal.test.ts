import { renderHook, waitFor } from '@testing-library/react';
import { PivotSortCriteria } from './sorting-utils';
import { usePivotTableDataOptionsInternal } from './use-pivot-table-data-options-internal';
import { mockPivotTableProps } from './__mocks__/mocks';

describe('usePivotTableDataOptionsInternal', () => {
  it('should correctly translate dataOptions into internalDataOptions', async () => {
    const { result } = renderHook(() =>
      usePivotTableDataOptionsInternal({ dataOptions: mockPivotTableProps.dataOptions }),
    );

    await waitFor(() => {
      expect(result.current.dataOptionsInternal).toBeDefined();
    });
    expect(result.current.dataOptionsInternal).toMatchSnapshot();
  });

  it('should udpate row (first) with new sorting', async () => {
    const sortCriteria = {
      rows: [
        {
          rowIndex: 0,
          sort: {
            direction: 'sortAsc',
            by: {
              valuesIndex: 0,
              columnsMembersPath: ['Female'],
            },
          },
        },
      ],
    } as PivotSortCriteria;
    const { result } = renderHook(() =>
      usePivotTableDataOptionsInternal({ dataOptions: mockPivotTableProps.dataOptions }),
    );

    result.current.updateSort(sortCriteria);

    await waitFor(() => {
      expect(result.current.dataOptionsInternal.rows![0].sortType).toBe(sortCriteria.rows[0].sort);
    });
  });

  it('should disable existing row (second) sorting', async () => {
    const sortCriteria = {
      rows: [
        {
          rowIndex: 1,
          sort: null,
        },
      ],
    };
    const { result } = renderHook(() =>
      usePivotTableDataOptionsInternal({ dataOptions: mockPivotTableProps.dataOptions }),
    );

    result.current.updateSort(sortCriteria);

    await waitFor(() => {
      expect(result.current.dataOptionsInternal.rows![1].sortType).toBeUndefined();
    });
  });
});
