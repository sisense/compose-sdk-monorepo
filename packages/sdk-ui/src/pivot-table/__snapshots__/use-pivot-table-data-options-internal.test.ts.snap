// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`usePivotTableDataOptionsInternal > should correctly translate dataOptions into internalDataOptions 1`] = `
{
  "columns": [
    {
      "__serializable": "DimensionalElement",
      "desc": "",
      "expression": "[Commerce.Gender]",
      "name": "Gender",
      "type": "text-attribute",
    },
  ],
  "grandTotals": {
    "columns": true,
    "rows": true,
    "title": "Grand Total",
  },
  "rows": [
    {
      "__serializable": "DimensionalElement",
      "desc": "",
      "expression": "[Category.Category]",
      "name": "Category",
      "type": "text-attribute",
    },
    {
      "__serializable": "DimensionalElement",
      "desc": "",
      "expression": "[Commerce.Age Range]",
      "name": "AgeRange",
      "type": "text-attribute",
    },
    {
      "__serializable": "DimensionalElement",
      "desc": "",
      "expression": "[Commerce.Condition]",
      "name": "Condition",
      "type": "text-attribute",
    },
  ],
  "values": [
    {
      "__serializable": "DimensionalElement",
      "aggregation": "sum",
      "attribute": {
        "__serializable": "DimensionalElement",
        "desc": "",
        "expression": "[Commerce.Cost]",
        "name": "Cost",
        "type": "numeric-attribute",
      },
      "desc": "",
      "name": "Total Cost",
      "sort": 0,
      "type": "basemeasure",
    },
    {
      "__serializable": "DimensionalElement",
      "aggregation": "sum",
      "attribute": {
        "__serializable": "DimensionalElement",
        "desc": "",
        "expression": "[Commerce.Revenue]",
        "name": "Revenue",
        "type": "numeric-attribute",
      },
      "desc": "",
      "name": "Total Revenue",
      "sort": 0,
      "type": "basemeasure",
    },
  ],
}
`;
